internals/overlord/logstate/loki/loki.go

```diff
@@ -0,0 +1,211 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package loki
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/internals/overlord/logstate/clienterr"
+	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+const maxRequestSize = 1000
+
+var requestTimeout = 10 * time.Second
+
+type Client struct {
+	target    *plan.LogTarget
+	remoteURL string
+	// buffered entries are "sharded" by service name
+	entries    map[string][]lokiEntry
+	numEntries int
+
+	httpClient http.Client
```

Comment:
I think this should probably be a `*http.Client` pointer. It probably doesn't matter right now, but if an `httpClient` was passed in it'd need to be a pointer to avoid copying. And for consistency with the examples in [net/http docs](https://pkg.go.dev/net/http).